include/swift/Parse/Parser.h

```diff
@@ -588,11 +588,22 @@ class Parser {
 
   /// When we have a token that is an identifier starting with '$',
   /// diagnose it if not permitted in this mode.
-  void diagnoseDollarIdentifier(const Token &tok) {
+  void diagnoseDollarIdentifier(const Token &tok,
+                                bool allowDollarIdentifier = false) {
     assert(tok.getText()[0] == '$');
 
-    if (tok.getText().size() == 1 ||
-        Context.LangOpts.EnableDollarIdentifiers ||
+    // If '$' is not guarded by backticks, offer
+    // to replace it with '`$`'.
+    if (Tok.getRawText() == "$") {
+      diagnose(Tok.getLoc(), diag::standalone_dollar_identifier)
+          .fixItReplace(Tok.getLoc(), "`$`");
+      return;
+    }
+
+    if (allowDollarIdentifier)
+      return;
+
+    if (tok.getText().size() == 1 || Context.LangOpts.EnableDollarIdentifiers ||
```

Comment:
I know this preserves the previous behavior, but shouldn't `LangOpts.EnableDollarIdentifiers` be prioritized over `allowDollarIdentifier`?

cc @rintaro

Reply:
These are OR conditions for allowing dollar identifiers. `EnableDollarIdentifiers == false` doesn't mean rejecting them everywhere. So the ordering shouldn't matter.